catalog/core/catalog-core-standardframework/src/main/java/ddf/catalog/impl/SourceOperations.java

```diff
@@ -0,0 +1,335 @@
+
+        SourceInfoResponse response;
+        SourceDescriptorImpl sourceDescriptor;
+        try {
+
+            // request
+            if (sourceInfoRequest == null) {
+                throw new IllegalArgumentException("SourceInfoRequest was null");
+            }
+
+            Set<SourceDescriptor> sourceDescriptors = new LinkedHashSet<>();
+            Set<String> ids = sourceInfoRequest.getSourceIds();
+
+            // Only return source descriptor information if this sourceId is
+            // specified
+            if (ids != null && !ids.isEmpty()) {
+                for (String id : ids) {
+                    if (!id.equals(this.getId())) {
+                        SourceUnavailableException sourceUnavailableException =
+                                new SourceUnavailableException("Unknown source: " + id);
+                        LOGGER.warn("Throwing SourceUnavailableException for unknown source: {}",
+                                id,
+                                sourceUnavailableException);
+                        throw sourceUnavailableException;
+
+                    }
+                }
+
+            }
+            // Fanout will only add one source descriptor with all the contents
+            Set<ContentType> contentTypes = frameworkProperties.getFederatedSources()
+                    .values()
+                    .stream()
+                    .filter(source -> source != null && source.isAvailable()
+                            && source.getContentTypes() != null)
+                    .map(Source::getContentTypes)
+                    .flatMap(Collection::stream)
+                    .collect(Collectors.toSet());
+
+            // only reveal this sourceDescriptor, not the federated sources
+            sourceDescriptor = new SourceDescriptorImpl(this.getId(), contentTypes);
+            sourceDescriptor.setVersion(this.getVersion());
+            sourceDescriptors.add(sourceDescriptor);
+
+            response = new SourceInfoResponseImpl(sourceInfoRequest, null, sourceDescriptors);
+
+        } catch (RuntimeException re) {
+            LOGGER.warn("Exception during runtime while performing create", re);
+            throw new SourceUnavailableException(
+                    "Exception during runtime while performing getSourceInfo",
+                    re);
+
+        }
+        return response;
+
+    }
+
+    /**
+     * Creates a {@link Set} of {@link SourceDescriptor} based on the incoming list of
+     * {@link Source}.
+     *
+     * @param sources {@link Collection} of {@link Source} to obtain descriptor information from
+     * @return new {@link Set} of {@link SourceDescriptor}
+     */
+    private Set<SourceDescriptor> getFederatedSourceDescriptors(Collection<FederatedSource> sources,
+            boolean addCatalogProviderDescriptor) {
+        SourceDescriptorImpl sourceDescriptor;
+        Set<SourceDescriptor> sourceDescriptors = new HashSet<>();
+        if (sources != null) {
```

Comment:
Clean up later. 268 onward could surely (hopefully, possibly, maybe..) be cleaned up by some good optionals or streaming.